File: backend/demo/audio_processor.py

import base64
import numpy as np

class AudioProcessor:
    @staticmethod
    def encode_audio(data, sample_rate):
        encoded = base64.b64encode(data.tobytes()).decode("utf-8")
        return {
            "realtimeInput": {
                "mediaChunks": [
                    {"mimeType": f"audio/pcm;rate={sample_rate}", "data": encoded}
                ]
            }
        }

    @staticmethod
    def process_audio_response(data):
        audio = base64.b64decode(data)
        return np.frombuffer(audio, dtype=np.int16)

File: frontend/lib/core/utils/audio_processor.dart
import 'dart:convert';
import 'dart:typed_data';

typedef JsonMap = Map<String, dynamic>;

class AudioProcessor {
  static JsonMap encodeAudio(Int16List data, {required int sampleRate}) {
    final b64 = base64Encode(Uint8List.view(data.buffer));
    return {
      'realtimeInput': {
        'mediaChunks': [
          {'mimeType': 'audio/pcm;rate=$sampleRate', 'data': b64},
        ],
      },
    };
  }

  static Int16List decodeAudio(String b64) {
    final bytes = base64Decode(b64);
    return Int16List.view(bytes.buffer);
  }
}
